### ISSUE TITLE
Tests fail on i686 and s390x architectures

### ISSUE THREAD

**Issue by ignatenkobrain** [nabijaczleweli/safe-transmute-rs#64]
# i686

```
---- error::unaligned::unaligned_slicing_integers stdout ----
thread 'error::unaligned::unaligned_slicing_integers' panicked at 'assertion failed: `(left == right)`
  left: `Ok([633890293370716671])`,
 right: `Err(Unaligned(UnalignedError { offset: 4, source: &[S] { len: 12 } }))`', tests/error/unaligned.rs:72:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
---- util::smoke_check_alignment_from_8 stdout ----
thread 'util::smoke_check_alignment_from_8' panicked at 'assertion failed: `(left == right)`
  left: `4`,
 right: `8`', tests/util/mod.rs:39:5
failures:
    error::unaligned::unaligned_slicing_integers
    util::smoke_check_alignment_from_8
```

# s390x

```
---- src/base.rs - base::from_bytes (line 38) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/base.rs - base::from_bytes_pedantic (line 77) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/base.rs - base::transmute_many (line 121) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/base.rs - base::transmute_many_mut (line 166) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/base.rs - base::transmute_many_permissive (line 208) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/full.rs - full::transmute_many (line 101) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/full.rs - full::transmute_many_mut (line 184) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/full.rs - full::transmute_many_pedantic (line 156) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/full.rs - full::transmute_many_pedantic_mut (line 240) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/full.rs - full::transmute_one_pedantic (line 72) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/full.rs - full::transmute_one (line 41) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/lib.rs - (line 101) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
4  | use std::intrinsics::size_of;
   |
4  | use std::mem::size_of;
   |
warning: trait objects without an explicit `dyn` are deprecated
 --> src/lib.rs:107:29
  |
7 | fn main() -> Result<(), Box<std::error::Error>> {
  |                             ^^^^^^^^^^^^^^^^^ help: use `dyn`: `dyn std::error::Error`
  |
  = note: `#[warn(bare_trait_objects)]` on by default
error: aborting due to 2 previous errors; 1 warning emitted
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/lib.rs - (line 126) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/lib.rs - (line 142) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/to_bytes.rs - to_bytes::transmute_one_to_bytes (line 238) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/to_bytes.rs - to_bytes::transmute_one_to_bytes_mut (line 280) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/to_bytes.rs - to_bytes::transmute_to_bytes (line 331) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/to_bytes.rs - to_bytes::transmute_to_bytes_many_unchecked (line 121) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/to_bytes.rs - to_bytes::transmute_to_bytes_many_unchecked_mut (line 175) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/to_bytes.rs - to_bytes::transmute_to_bytes_unchecked (line 19) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/to_bytes.rs - to_bytes::transmute_to_bytes_unchecked_mut (line 69) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/trivial.rs - trivial::transmute_trivial (line 176) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/trivial.rs - trivial::transmute_trivial_many (line 249) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/trivial.rs - trivial::transmute_trivial_many_mut (line 286) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- src/trivial.rs - trivial::transmute_trivial_pedantic (line 214) stdout ----
error[E0433]: failed to resolve: maybe a missing crate `core`?
  --> src/../tests/test_util/le_to_native.rs:14:13
   |
14 |         use core::mem::size_of;
   |             ^^^^ maybe a missing crate `core`?
error[E0425]: cannot find function `size_of` in this scope
  --> src/../tests/test_util/le_to_native.rs:16:37
   |
16 |         for elem in self.chunks_mut(size_of::<T>()) {
   |                                     ^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
3  | use std::intrinsics::size_of;
   |
3  | use std::mem::size_of;
   |
error: aborting due to 2 previous errors
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
```

**Comment by nabijaczleweli** [comment]
Fixed the i686 bit (`master` branch, at least 09acd5012fb5ab1f84793fd79010111c0614b19d). I don't have s390x hardware to test on, unfortunately. Do you have a recommended virtualisation solution?

**Comment by ignatenkobrain** [comment]
I think qemu can emulate s390x... However, if you have patches - I can easily test them on Fedora machines.

**Comment by nabijaczleweli** [comment]
Nothing immediately sprung to mind, so I'll have to play around with this. After some dicking about, QEMU appears to work, though, thanks for the pointer.

For future reference, [stretch](http://ftp.nl.debian.org/debian/dists/stretcb/main/installer-s390x/current/images/generic/) images, 
```
qemu-system-s390x -machine s390-ccw-virtio -cpu max,zpci=on -smp 6 -nographic -m 4g \
   -nic user,model=virtio,hostfwd=tcp::2222-:22 \
   -kernel kernel.debian -initrd initrd.debian -append 'BOOT_DEBUG=3' \
   -drive file=/dev/zvol/filling/store/nabijaczleweli/vm-s390x,if=none,id=drive-virtio-disk0,format=raw,cache=none \
   -device virtio-blk-ccw,devno=fe.0.0001,drive=drive-virtio-disk0,id=virtio-disk0,bootindex=1,scsi=off
```
and `mknod /dev/ttyS1 c 4 65` in the second shell worked.

**Comment by nabijaczleweli** [comment]
Okay, I figured it out I think. @ignatenkobrain mind verifying on your set-up (`master` branch, at least 057931a23883967cbcb06c85d7ce2a7734fbc4a1)?

**Comment by ignatenkobrain** [comment]
@nabijaczleweli yep, applying both fixes help :)

**Comment by nabijaczleweli** [comment]
Ace

**Comment by nabijaczleweli** [comment]
Released in [`v0.11.1`](https://github.com/nabijaczleweli/safe-transmute-rs/releases/tag/v0.11.1).
